pandaserver/taskbuffer/OraDBProxy.py

```diff
@@ -22831,54 +22831,70 @@ def reportWorkerStats_jobtype(self, harvesterID, siteName, paramsList):
             return False, "database error"
 
     # get stat of workers
-    def getWorkerStats(self, siteName):
+    def getWorkerStats(self):
         comment = " /* DBProxy.getWorkerStats */"
         methodName = comment.split(" ")[-2].split(".")[-1]
-        tmpLog = LogWrapper(_logger, methodName + f" < siteName={siteName} >")
+        tmpLog = LogWrapper(_logger)
         tmpLog.debug("start")
         try:
             # set autocommit on
             self.conn.begin()
             # sql to get nPilot
-            sqlP = "SELECT getJob+updateJob FROM ATLAS_PANDAMETA.SiteData "
-            sqlP += "WHERE HOURS=:hours AND FLAG IN (:flag1,:flag2) "
+            # sqlP = ("SELECT getJob+updateJob FROM ATLAS_PANDAMETA.SiteData "
+            #         "WHERE HOURS=:hours AND FLAG IN (:flag1,:flag2) ")
+            # varMap = dict()
+            # varMap[":hours"] = 1
+            # varMap[":flag1"] = "production"
+            # varMap[":flag2"] = "analysis"
+            # self.cur.execute(sqlP + comment, varMap)
+            # res = self.cur.fetchone()
+            # if res is not None:
+            #     (nPilot,) = res
+            # else:
+            #     nPilot = 0
+            # sql to get stat of active workers
+            sqlGA = (
+                "SELECT SUM(n_workers), computingSite, harvester_ID, jobType, resourceType, status "
+                "FROM ATLAS_PANDA.Harvester_Worker_Stats "
+                "WHERE lastUpdate>=:time_limit "
+                "GROUP BY computingSite,harvester_ID,jobType,resourceType,status "
+            )
             varMap = dict()
-            varMap[":hours"] = 1
-            varMap[":flag1"] = "production"
-            varMap[":flag2"] = "analysis"
-            self.cur.execute(sqlP + comment, varMap)
-            res = self.cur.fetchone()
-            if res is not None:
-                (nPilot,) = res
-            else:
-                nPilot = 0
-            # sql to get stat
-            sqlG = "SELECT SUM(n_workers), COUNT(harvester_ID), jobType, resourceType, status FROM ATLAS_PANDA.Harvester_Worker_Stats "
-            sqlG += "WHERE computingSite=:siteName "
-            sqlG += "GROUP BY resourceType,status "
+            varMap[":time_limit"] = datetime.datetime.utcnow() - datetime.timedelta(hours=4)
+            self.cur.execute(sqlGA + comment, varMap)
+            res_active = self.cur.fetchall()
+            # sql to get stat of finished workers from wokrer table
+            sqlGF = (
+                "SELECT COUNT(*), computingSite, harvesterID, jobType, resourceType, status "
+                "FROM ATLAS_PANDA.Harvester_Workers "
```

Comment:
How often is this function called? If it hits the Harvester_Workers table frequently it would be better to change harvester to report # of finished workers as well to Harvester_Worker_Stats.

Reply:
It's triggered by harvester cacher, so typically every 10 minutes /  n_harvester_instances.

> finished workers as well to Harvester_Worker_Stats.
That will be ideal, but I may need to check all functions elsewhere related Harvester_Worker_Stats to make sure they won't break with finished workers added

Reply:
I let Harvester report to Harvester_Worker_Stats (https://github.com/HSF/harvester/pull/205) so no need to query Harvester_Workers.